### PR TITLE
Fix ReadBufferState dispose cases

### PR DESCRIPTION
**Pull request by MaxPatri**
I moved the initialization of ReadBufferState after other related operations to prevent skipping Dispose in case of an exception, and consequently, to avoid memory leaks.

Found by Linux Verification Center (linuxtesting.org) with SVACE.

### CODE REVIEW

**Comment by stephentoub** [comment]
> to avoid memory leaks

What leaks?

**Comment by huoyaoyuan** [comment]
If `ReadStack.Initialize` throws, the array rented in `ReadBufferState` won't be returned. `new JsonReaderState` won't throw because it's pure POCO initialization.
There won't be leak, because failing to return pooled array isn't a fatal leak.

**Comment by eiriktsarpalis** [comment]
This is the code for `ReadStack.Initialize`:

https://github.com/dotnet/runtime/blob/db9a4c6ecf40c49ebfa253cef92fdca683982fbf/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs#L96-L110

And this is the code for the `JsonReaderState`:

https://github.com/dotnet/runtime/blob/db9a4c6ecf40c49ebfa253cef92fdca683982fbf/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderState.cs#L39-L54

Excluding potential bugs in the underlying implementation(s), neither of these calls are expected to throw. Is there a real-world scenario that could result in unreturned buffers?

**Comment by MaxPatri** [comment]
Moved ReadBufferState initialization under try clause.

**Comment by eiriktsarpalis** [comment]
> Moved ReadBufferState initialization under try clause.

That's not quite that I meant. My suggestion was to either keep the current order (but nest the initialization methods inside the `try` clause) or add a comment to your original change explaining why the `ReadBufferState` is being initialized last.

**Comment by MaxPatri** [comment]
Initial change was returned with comment

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs [hunk 1]
```diff
@@ -553,10 +553,11 @@ public static partial class JsonSerializer
             {
                 Debug.Assert(listTypeInfo.IsConfigured);
 
-                ReadBufferState bufferState = new(listTypeInfo.Options.DefaultBufferSize);
                 ReadStack readStack = default;
                 readStack.Initialize(listTypeInfo, supportContinuation: true);
                 JsonReaderState jsonReaderState = new(readerOptions);
+                // bufferState is initialized last to exlcude any exceptions before entering the try/finally block where bufferState.Dispose call is executed
```
Comment (jnyrup) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs:559]: exlcude -> exclude

Comment (MaxPatri) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs]: I'm sorry, fixed

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs [hunk 2]
```diff
@@ -29,11 +29,11 @@ public partial class JsonTypeInfo<T>
             ReadStack readStack = default;
             readStack.Initialize(this, supportContinuation: true);
             var jsonReaderState = new JsonReaderState(options.GetReaderOptions());
-            ReadBufferState bufferState = null;
+            // bufferState is initialized last to exlcude any exceptions before entering the try/finally block where bufferState.Dispose call is executed
+            var bufferState = new ReadBufferState(options.DefaultBufferSize);
 
             try
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs]: A terser message and tactical use of indentation should suffice to convey the message:
```suggestion

            // Note: ReadBufferState rents pooled buffers.
            ReadBufferState bufferState = new(options.DefaultBufferSize);
            try
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs [hunk 3]
```diff
@@ -62,11 +62,11 @@ public partial class JsonTypeInfo<T>
...
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs]:
```suggestion

            // Note: The ReadBufferState ctor rents pooled buffers.
            ReadBufferState bufferState = new(options.DefaultBufferSize);
            try
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs (hunk 2 above)
Comment (MaxPatri) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs]: Fixed

**Review by eiriktsarpalis — APPROVED** [review]
Thanks

**Comment by MaxPatri** [comment]
Is this normal for check "runtime (Build browser-wasm linux Release LibraryTests)" to be running so long? Or may be it is stuck?
Also I would like to know what should I do with failed checks? I don't see restart button or something else and I assume that my changes could not affect the build or tests.

**Pull request merged**